### ISSUE TITLE
Use cases for synchronous subscription.

### ISSUE THREAD

**Issue by benjamingr** [zenparsing/es-observable#38]
It would be great if we had a list of use cases for synchronous subscription in order to better understand if/when `[Symbol.observer]` is needed.

Please include:

 - **Description** - Why is this use case important and who are the users. 
 - **Code** - a minimal example demonstrating the use case and showcasing it. preferably codepen/jsbin/jsfiddle but a gist is also fine, demonstrating the issue. 
 - **Why synchronous subscription?** - why would asynchronous subscription not work here, what benefit does synchronous subscription bring to the table?
 

Pinging people who want synchronous inspection or have expressed interest in it before: @jhusain @blesh @tgriesser 

Bonus points for weakness of async subscription compared to existing DOM or Node/io APIs. If you use "performance" as a reason make sure you understand what an async trampoline means (batching asynchronous function calls), how async tagging works (only defer actions that were not already async and only once per chain) and so on.

**Comment by blesh** [comment]
Basically all use cases for Observable when you want Observable to be performant.

If you're programming functionally, you should be using `[Symbol.observer]` every single time, otherwise you're wasting time scheduling things on a micro task queue that don't need to be scheduled.

**Comment by benjamingr** [comment]
> otherwise you're wasting time scheduling things on a micro task queue that don't need to be scheduled.

Repeating myself:

>  you use "performance" as a reason make sure you understand what an async trampoline means (batching asynchronous function calls), how async tagging works (only defer actions that were not already async and only once per chain) and so on.

Basically, you don't really need to put almost anything on the microtask queue. I see my previous points about the trampoline and async tagging were unclear. Allow me to elaborate.

###Trampoline:

You have:

```js
for(var i = 0; i < 1000; i++){
    observables[i].subscribe(funcs[i]);
}
```

Let's say all 1000 observables have results for all 1000 functions (the "worst case").

Naively, one would schedule 1000 microtasks and perform a thousand calls to the system. In practice one schedules a single function on the microtask queue that calls all 1000 callbacks in sequence at once. So you pay for the latency of the scheduler exactly once and not 1000 times. This is utilized for example in [Bluebird's scheduling](https://github.com/petkaantonov/bluebird/blob/master/src/async.js). A queue is created and function calls are batched. 

###Async tagging

Although the trampoline makes things generally pretty fast, you still have to pay latency once per batch. This is wasteful. Instead, remember that _we_ created the observable, in our end it was something like:

```js
function listen(element, eventName) {
    return new Observable(sink => {
        // call sink.next
    });
}
```

But on the Observable side _we_ get to implement `next` so we:

 - Create the sink and a boolean flag, start the flag as false.
 - Call the constructor executor argument.
 - Set the flag to true.
 - When `next` is called, if the flag is false set it to true and schedule a microtask via the trampoline. Otherwise, if the flag isn't false (that is, the constructor ran to completion before `next` was called) - call the function synchronously.

This means that in 99.99% of cases `fn()` is called synchronously, and in the 0.01% of cases it's not _it's in cases where it is actively saving you from Zalgo_. 

In purely functional coding - you should not notice any performance impact. You can see this [here in bluebird](https://github.com/petkaantonov/bluebird/blob/20915380ce3162637ffce65e362bef11e27d6a89/src/promise.js#L481-L487) here. 

As for async stack traces and stack construction that @jhusain mentioned it is taken care of by the platform anyway (async stack traces) so it should not be an issue.

**Comment by blesh** [comment]
>  So you pay for the latency of the scheduler exactly once and not 1000 times.

In that one non-real-world case, yes, you've done only 1 unnecessary thing. But performance is more about macro performance than micro performance.

In an application that is written completely functionally with observables, you're going to incur that cost for every flatMapped/merged/cattered observable in your application, throughout the life span of your app. 

As I'm sure @jhusain and @trxcllnt will be able to tell you, Netflix has need of Observables that will run in non-JITTed environments on weak hardware. In that scenario, you'll likely feel scheduling. Now arguably, "native" Observables won't make it to those platforms any time soon; But any library that wants to target this spec will be less ergonomic to use in those environments because of this feature.

I don't deny the need for an async subscribe for the masses. I just wish there was a more ergonomic way for people that know what they're doing to use the sync subscribe.

**Comment by blesh** [comment]
> I don't deny the need for an async subscribe for the masses.

... but I will spend my career explaining to people why they don't need to use it.

**Comment by tgriesser** [comment]
@benjamingr I'll try to gather some thoughts on this, as I've been using `[Symbol.observe]` with RxJS Next and too have found myself feeling the same re: ergonomics.

Just so I can better understand the desire for the async subscription here, do you think you could also provide a similar answer the inverse of your question:

It would be great if we had a list of use cases for *asynchronous* subscription in order to better understand if/when *Zalgo occurs with Observables*.

**Comment by blesh** [comment]
@tgriesser it's mostly around problems with mutating state and imperative programming.

```js
var foo = true;
myObservable.subscribe(x => {
  foo = false;
});
assert(foo);
```

People write (bad) code assuming that the `subscribe` method will be asynchronous all the time, and they get bitten.  This gets exacerbated when you're dealing with a third party library that returns observables:

```js
var myObservable = thirdParty.getDataSource();

myObservable.subscribe(x => doThings(x));

notifySubscriptionToDataSource();
```

Now imagine the above `thirdParty.getDataSource` was created in one of two ways:

```js
var thirdParty = {
  getDataSource() {
    return Rx.Observable.range(0, 100);
  }
}
```

or 

```js
var thirdParty = {
  getDataSource() {
    return Rx.Observable.create(observer => {
       var i = 0;
       var id = setInterval(() => observer.next(i++), 1000);

       return () => {
         clearInterval(id);
       }
    });
  }
};
```

In the former case, `notifySubscriptionToDataSource` is not called until *after* the subscription has started and the entire observable has completed. In the latter case, `notifySubscriptionToDataSource` is called *before* the subscription has started.

**Comment by trxcllnt** [comment]
>This means that in 99.99% of cases fn() is called synchronously, and in the 0.01% of cases it's not it's in cases where it is actively saving you from Zalgo.

The Observable monad is the excalibur that slays Zalgo.

>In purely functional coding - you should not notice any performance impact.

If your code is purely functional, it doesn't matter whether subscribe is synchronous or asynchronous, in which case any asynchronous scheduling is wasted effort.

**Comment by zenparsing** [comment]
Thanks @benjamingr for the examples.

@jhusain 's comments on the DOM in the other thread got me thinking about something about our current async-subscription design.

Keep in mind that in the current design, `subscribe` is async but `next` is sync.

```js
// This is a multicast observable:  you can put data into it and it will send that data to
// all subscribers
let multicast = createMulticaster();

// We add a subscription to the multicast source
let subscription = multicast.subscribe(value => console.log(`Got value: ${ value }`));

// We tell the multicast source to send a data value to all subscribers
multicast.send("some-data");
```

The question is:  should the onNext callback above be executed?  Should "Got value: some-data" be logged?  What is the intuitive answer?

**Comment by trxcllnt** [comment]
@zenparsing yes, "Got value" should be logged. The root of this problem is a fundamental disagreement between imperative and functional programming. By introducing a Subject (an Observable + Observer), you've gone off the reservation of immutable determinism and out into the sand dunes of race conditions.

Observables aren't fancy EventEmitters, and we shouldn't proscribe behaviors that don't make sense in purely functional code.

**Comment by zenparsing** [comment]
@trxcllnt Cool.  (Although we can't make the ES standard library into an imperative vs. functional battleground.  Instead, we should focus on making the API consistent, simple, and usable.)

Anyone else have an intuition?

**Comment by blesh** [comment]
> we should focus on making the API consistent, simple, and usable

I agree. It's *my opinion* that it's "consistent" if it's synchronous if it can be. It might even force people to learn better programming practices. Keep in mind, if this is implemented, we're only introducing a *type* nothing in the JS core would return this type, so using Observable would be a strictly *opt-in* programming experience. Furthermore, those that want that experience are probably looking to program functionally rather than imperatively.

On a related note, if "simple" is a goal, then I *personally* think we shouldn't try to get the disposal behavior to work with generators, because the behavior is decidedly less simple than prior art (RxJS 2).

**Comment by domenic** [comment]
I think if people are going to opt-in to functional programming, they should use a library, and we should not standardize this in the JS language, which is pretty clearly imperative.

**Comment by domenic** [comment]
@zenparsing I have a hard time building intuition about your example because a "multicast observable" and "send" are not things that appear in the proposal.

**Comment by trxcllnt** [comment]
@domenic That's the point -- you can opt-in to mutability, but you can't opt-in to immutability.

If someone wants asynchronous subscriptions, it's very simple to schedule the subscription to happen asynchronously: `Observable.subscribeOn(Scheduler.nextTick)`.

**Comment by blesh** [comment]
>we should not standardize this in the JS language, which is pretty clearly imperative

You know, we should remove first-class functions then, right? This is more FUDD, frankly.

**Comment by jhusain** [comment]
I think we should probably keep this thread focused on use cases for asynchronous subscription.

In the other thread I outlined two: 

1. modeling EventTarget which allows for synchronous notification.
2. epoll-style push/pull IO in which a notification is sent saying a value is available, but the value is retrieved via a pull. This allows the potentially  expensive process of retrieving a value to be avoided if the consumer is not interested in the data.

**Comment by jhusain** [comment]
FYI I will try and follow up with code examples for each of there use cases when I get a chance.

**Comment by trxcllnt** [comment]
3. Depth-first vs. breadth-first `expand` is only possible with synchronous subscribe.

**Comment by domenic** [comment]
The epoll case doesn't make any sense, as that's definitely not how epoll works. (epoll includes a buffer, for one, and when not used directly in a polling fashion, it's commonly wrapped in an event loop library, which will make it async.) https://blog.domenic.me/reading-from-sockets/ may be helpful.

**Comment by domenic** [comment]
@trxcllnt please follow the requirements outlined in the OP, your one-sentence description is pretty impenetrable.

**Comment by jhusain** [comment]
I would be very interested if anyone on this thread Who is advocating for synchronous subscription could come up with a use case in which asynchronous subscription meaningfully hurts performance. Thus far I have been a unable to prove this in a real-world use case. 

It's important to understand that the committee is strongly swayed by real world use cases. Theoretical arguments, while they may be valid, are going to be less influential.

@trxcllnt any ideas? You have a lot of experience using observable to build very chatty event streams (ex. Gestures, scrolling virtualization). Can you demonstrate that asynchronous subscription will hurt performance in any of these use cases.

JH

**Comment by trxcllnt** [comment]
@domenic @jhusain [Example](http://jsbin.com/jidoqanija/edit?js,console) of a recursive tree-walk as depth-first, breadth-first, and asynchronous breadth-first parameterized by scheduler.

```javascript
var Observable = Rx.Observable;

Observable.prototype.expand = function (selector) {
  var source = this;
  return Observable.create(function (observer) {

    var disposables = new Rx.CompositeDisposable();

    subscribeInner(source);

    return disposables;

    function subscribeInner(innerObs) {
      var innerSad = new Rx.SingleAssignmentDisposable();
      disposables.add(innerSad);
      innerSad.setDisposable(innerObs.subscribe(
        function onNext(x) {
          subscribeInner(selector(x));
          observer.onNext(x);
        },
        function onError(e) {
          observer.onError(e);
        },
        function onCompleted() {
          disposables.remove(innerSad);
          if(disposables.length === 0) {
            observer.onCompleted();
          }
        }
      ));
    }
  });
};

var fileSystem = {
  "C:": {
    "ProgramFiles": {
      "Chrome": {
        "app": "chrome.exe",
        "images": {
          "icon": "chrome-icon.ico"
        }
      },
      "Firefox": {
        "app": "firefox.exe",
        "images": {
          "icon": "firefox-icon.ico",
          "user": {
            "ptaylor": {
              "icon": "ptaylor-icon.png"
            }
          }
        }
      },
      "Opera": {
        "app": "Opera.exe",
        "images": {
          "icon": "Opera-icon.ico"
        }
      }
    }
  }
};

function ls(path, tree, scheduler) {
  scheduler = scheduler || (Rx.Scheduler.currentThread);
  if(path) {
    path = path + "/";
  }
  return Observable.create(function (observer) {
    return Observable
      .fromArray(Object.keys(tree), scheduler)
      .map(function(key) {
        return { path: path + key, value: tree[key] };
      })
      .subscribe(observer);
  });
}

function listResources(scheduler) {
  ls("", fileSystem, scheduler)
    .expand(function (resource) {
      var path = resource.path;
      var fileOrDir = resource.value;
      if(typeof fileOrDir !== "object") {
        return Observable.empty();
      }
      return ls(path, fileOrDir, scheduler);
    })
    .filter(function(resource) {
      return typeof resource.value !== "object";
    })
    .subscribe(function(resource) {
      console.log(resource.path + " -- " + resource.value);
    });
}

console.log("Depth-first:");
listResources(Rx.Scheduler.immediate);

console.log("Breadth-first:");
listResources(Rx.Scheduler.currentThread);

console.log("Asynchronous breadth-first:");
listResources(Rx.Scheduler.timeout);

/* output:
Depth-first:
C:/ProgramFiles/Chrome/app -- chrome.exe
C:/ProgramFiles/Chrome/images/icon -- chrome-icon.ico
C:/ProgramFiles/Firefox/app -- firefox.exe
C:/ProgramFiles/Firefox/images/icon -- firefox-icon.ico
C:/ProgramFiles/Firefox/images/user/ptaylor/icon -- ptaylor-icon.png
C:/ProgramFiles/Opera/app -- Opera.exe
C:/ProgramFiles/Opera/images/icon -- Opera-icon.ico
Breadth-first:
C:/ProgramFiles/Chrome/app -- chrome.exe
C:/ProgramFiles/Firefox/app -- firefox.exe
C:/ProgramFiles/Chrome/images/icon -- chrome-icon.ico
C:/ProgramFiles/Opera/app -- Opera.exe
C:/ProgramFiles/Firefox/images/icon -- firefox-icon.ico
C:/ProgramFiles/Opera/images/icon -- Opera-icon.ico
C:/ProgramFiles/Firefox/images/user/ptaylor/icon -- ptaylor-icon.png
Asynchronous breadth-first:
C:/ProgramFiles/Chrome/app -- chrome.exe
C:/ProgramFiles/Firefox/app -- firefox.exe
C:/ProgramFiles/Chrome/images/icon -- chrome-icon.ico
C:/ProgramFiles/Opera/app -- Opera.exe
C:/ProgramFiles/Firefox/images/icon -- firefox-icon.ico
C:/ProgramFiles/Opera/images/icon -- Opera-icon.ico
C:/ProgramFiles/Firefox/images/user/ptaylor/icon -- ptaylor-icon.png
*/
```

**Comment by benjamingr** [comment]
@tgriesser 

> Just so I can better understand the desire for the async subscription here, do you think you could also provide a similar answer the inverse of your question:

Sure: 

```js
someExternalSource().subscribe(val => showUi(val));
hideUi();
```

If `someExternalSource` suddenly starts fetching the result from cache at the second page load - we have a hidden UI instead of a visible one with the value because the subscription ran synchronously "this one time". These bugs are very hard to find and I think we should do our best to help prevent them.

----------

@blesh 

> In that one non-real-world case, yes, you've done only 1 unnecessary thing. But performance is more about macro performance than micro performance.

Actually, you'd do zero async scheduling in the vast majority of examples. In the vast majority of examples your source is actually asynchronous already so you'll not need any scheduling because of async tagging anyway.

> Keep in mind, if this is implemented, we're only introducing a type; Nothing in the JS core would return this type, so using Observable would be a strictly opt-in programming experience.

The idea is that the JS host environments would return observables for multiple valued objects. Otherwise - why would we even be specifying observables? People can just use third party libraries. 

@trxcllnt 

> If your code is purely functional, it doesn't matter whether subscribe is synchronous or asynchronous, in which case any asynchronous scheduling is wasted effort.

Right, but as I said before in my message about async scheduling, you're only doing it if you're running it through a sync data source. 

> <code example for `expand`>

I'm not sure what this demonstrates at all.
 
> If someone wants asynchronous subscriptions, it's very simple to schedule the subscription to happen asynchronously: Observable.subscribeOn(Scheduler.nextTick).

For what it's worth, I'm very amendable to allowing a `setScheduler` for letting people choose how to schedule their observables. Although that greatly complicates the proposal it's very useful for a variety of other use cases too. 

@zenparsing 

> The question is: should the onNext callback above be executed? Should "Got value: some-data" be logged? What is the intuitive answer?

Of course, 100% yes. 

@jhusain

> FYI I will try and follow up with code examples for each of there use cases when I get a chance.

Please do, that's why I opened this thread in the first place.

> I would be very interested if anyone on this thread Who is advocating for synchronous subscription could come up with a use case in which asynchronous subscription meaningfully hurts performance. Thus far I have been a unable to prove this in a real-world use case.

Yes, that's what I've been seeing too. A lot of big words and appealing to 'pure functional programming' without any such data. Data would go a long way.

**Comment by benjamingr** [comment]
Synchronous subscription in terms of an API is a huge footgun for anyone not doing a very particular type of "pure functional" programming where not only data is only moved in Observables, the UI layer doesn't interact with subscriptions directly. For those people, which I believe are the majority of users, we're introducing a class of concurrency bugs which are _extremely hard to debug_. 

Functional programming doesn't mean we don't have state, it's just controlled. Doing monadic pure state book-keeping has proven to be _very unpopular_ although _widely available_ in JavaScript and catering to people doing that exact type of observable usage over the vast majority of users sounds like something the TC will never approve. Especially if async subscription allows for the same expressiveness and as @trxcllnt said, you're not even aware of async subscription when doing async scheduling. 

What the TC _might_ approve is a synchronous subscription API _if it is proven useful_ and you can _do things with it you can't with async subscription_. Which is why what I've asked for from the start is _code examples with detailed explanation that necessitate sync subscription. So far we didn't get any.

**Comment by jhusain** [comment]
Thanks for the code example @trxcllnt. Can you provide a little more context around it? I can see that using an immediate way scheduler vs a trampoline scheduler (the poorly-named currentThread scheduler) easily switches from a depth-first search to a breadth-first search. Would I not get the same result from a using a version of the trampoline scheduler with a stack vs a queue? It's early in the morning so I'm sorry if I'm misunderstanding the example. 

. 

JH

**Comment by jhusain** [comment]
The most obvious problem with forced asynchronous subscription is the inability to faithfully model EventTarget. This is a serious issue, because it is one of the stated goals of the proposal. This is a worthwhile goal because EventTarget is both ubiquitous and awful. Observable would be a big improvement over EventTarget because addEventListener/removeEventListener methods hanging off an object cannot be ergonomically composed. Ergonomic composition of push streams is a big win, sync subscription or not.

Using Observable to model EventTarget is straightforward with synchronous subscription.

Let's say we had the following contract:

EventTargetObservable.listen(eventName, useCapture): Observable<Event>

Anything that currently implements EventTarget could easily opt into this contract without breaking changes. Now take a look at this code example which demonstrates the hazard of asynchronous subscription.

// receives notification
domElement.addEventListener("click", e => e.preventDefault(), false);

// receives notification
domElement.listen("click", false)[Symbol.observer](e => e.preventDefault());

// misses data!
domElement.listen("click", false).subscribe(e => e.preventDefault());

domElement.dispatchEvent("click", new ClickEvent());

Any proposal that attempts to model Event target will run into the same issue if it attempts to schedule subscription on a separate job. Note that scheduling notification is also not an option, because preventDefault() must be invoked within the same job.

JH

**Comment by benjamingr** [comment]
@jhusain 

So just to be clear, we are arguing that `EventTarget` requires synchronous subscription because __it would be impossible to interop between `addEventListener` and Observables otherwise__, right?

This is the sort of use case I was talking about, nice work!

I've written a concrete example of this sort of issue, if this is what you've had in mind let me know and I'll formalize it in a markdown document:

HTML:

```js
<div id="father">
    <div id="child">Click</div>
</div>
```

JavaScript:

```js
let father = document.querySelector("#father"),
    child  = document.querySelector("#child");

father.addEventListener("click", e => alert("Hi"), false);
child.listen("click", false).subscribe(e => e.stopPropagation());

child.click();

```

Result: the `alert` fires since addEventListener subscribes synchronously but `subscribe` subscribes asynchronously.

Implication:  It is impossible to stop propagation or prevent default behaviour of _synchronous_ events that are listened for via _addEventListener_ handlers.

Note that in asynchronous events or events dispatched by the platform would not be an issue at all, we are strictly talking about events dispatched by JavaScript here.

Also note that `preventDefault` doesn't work _except for `click` events_, as it would not be a [trusted event](http://www.w3.org/TR/DOM-Level-3-Events/#trusted-events), propagation would still work. 

Let me know if this is what you've had in mind.

**Comment by zenparsing** [comment]
> Implication: It is impossible to stop propagation or prevent default behaviour of synchronous events that are listened for via addEventListener handlers.

Clarification:  it's not really a propagation or prevent-default issue (those are symptoms, rather than cause).  The root issue (which I alluded to upstream) is that **we expect any events which occur after a call to subscribe to be picked up by that subscription**.

In other words, we intuitively want *subscription* to happen synchronously.  I think this is true for all observables, and my current opinion is that async subscription just won't work for that reason.

This is what I'm thinking:

- Subscription should happen synchronously, as pointed out above.
- In order to protect against zalgo, "well-behaved" observables **should never** execute any observer callbacks synchronously.

I'm not quite sure what this means for the API, at the moment.

**Comment by benjamingr** [comment]
> Clarification: it's not really a propagation or prevent-default issue (those are symptoms, rather than cause). The root issue (which I alluded to upstream) is that we expect any events which occur after a call to subscribe to be picked up by that subscription.

We can just queue those `next` calls and dispatch them on the next event loop iteration. This is what the browser does anyway in asynchronous events. The issue I describe above is specifically an issue of DOM events being dispatched by JavaScript synchronously and is an artifact of the current API for event dispatching - not the model.
 
If the DOM events were dispatched from JS asynchronously, this would not be an issue. What I believe @jhusain is speaking of is specifically compatibility with the way events currently work in the DOM.

**Comment by zenparsing** [comment]
@benjamingr There's a general problem here that I think is important.  Any pub-sub mechanism which allows synchronous dispatch won't work in the expected way if implemented with async subscription.  In any such system, you have to register the subscriber synchronously.

**Comment by benjamingr** [comment]
@zenparsing I don't understand (I updated my above comment, maybe that didn't show). Would you mind explaining why? Promises allow synchronous dispatch and the subscriber is _effectively_ dispatched asynchronously. This is true for a wide variety of pub/sub mechanisms in programming language. Am I missing something here?

**Comment by zenparsing** [comment]
@benjamingr 

We might have a bit of terminology confusion.  Let's start with promises.

Promises allow synchronous *resolution*, but only async *notification*.

Let's leave aside whether observable subscription is async or not.  For observables, *notification* is synchronous.

```js
let obs = new Observable((next) => {
    // After 1 second, notify the observer
    window.setTimeout(_=> {
        console.log("before notification");
        next(123);
        console.log("after notification");
    }, 1000);
});

obs.subscribe(value => {
    console.log("notification received: " + value);
});
```

After the timeout, we'll see the following log output:

```js
before notification
notification received: 123
after notification
```

Now let's say that we have a pub-sub thing:

```js
Pub.prototype.observe = function() {
  return new Observable(next => {
    // The "subscriber function"
    this.listeners.add(next);
    return _=> this.listeners.delete(next);
});

Pub.prototype.dispatch = function(event) {
  // Ignoring error handling here...
  this.listeners.forEach(next => next(event));
};

let publisher = new Pub();

// Let's create a subscription
publisher.observe().subscribe(value => console.log("received value: " + value));

// Now let's dispatch an event
publisher.dispatch("something");
```

If the subscriber function runs asynchronously, then the subscriber won't receive the event above because the event is dispatched before the subscriber function is run.  I think that's very surprising.

**Comment by benjamingr** [comment]
@zenparsing

> If the subscriber function runs asynchronously, then the subscriber won't receive the event above because the event is dispatched before the subscriber function is run. I think that's very surprising.

That's a good point, isn't this only the case if subscriptions don't buffer values pushed? (even if they do buffer, they only need to do so when asynchronousity is not a part of any part of the chain because of async tagging).

What @jhusain has pointed out is that this might require _unbounded_ buffering which can be an issue if we can come up with a realistic use case where unbounded buffering will happen in practice) . (This is in _addition_ to, and is separate from the EventTarget compatibility issue he has raised above)

**Comment by jhusain** [comment]
@zenparsing Note that scheduling notifications (ie onNext, onError, and onCompleted) can not be done asynchronously either. This would break preventDefault which needs to be called within the same job as the event notification. _It's important to understand that neither subscription nor notification can be forced to occur asynchronously if Observable is to faithfully model EventTarget_.

Is this a serious problem? No. When it comes to streams, there is no one right answer to whether subscriptions or notifications should happen sync/async. That's why Observable is flexible enough to support either use case. If there is a risk of data loss, producers can either choose to asynchronously schedule notifications.  Alternately they can force consumers to subscribe synchronously and work harder to maintain their invariants. With great power comes great responsibility.

This is a tradeoff. It's not a problem that we can fix with a type. With Promises there is a clear right solution. Enforcing asynchronous notification and buffering is very inexpensive because there is only one value. Neither of these things are necessarily true for streams. 

As an example, a multicast producer (like EventTarget) could choose to schedule notifications asynchronously _or_ allow sync subscription and make a copy of the observer list before notifying. Each would have the desired effect of avoiding plan interference, like an observer causing a change to the producer's observer list as notifications were being broadcast. Each would also have different space/time tradeoffs. Producers can make an informed choice based on the predicted frequency of events. _We (the committee) cannot make the right decision inside of the Observable library._ We can only focus on giving implementers the options to make the right tradeoff. That's why Observable was always conceived of as being fully synchronous, but allowing producers or consumers to opt into async subscription or async notification.

JH

**Comment by blesh** [comment]
Just so I have this straight, the problem with async `subscribe` is basically this:

```js
Observable.fromEvent(document, 'click').subscribe(x => console.log('clicked'));
document.dispatchEvent(new Event('click'));
```

The above code should log "clicked", but with the current es-observable spec, it will not.

This is because:

1. EventTarget.addEventListener registers the event synchronously ... and...
2. EventTarget.dispatchEvent fires the event synchronously.

And this holds true for all DOM events?

Meaning in the current design, it would be advisable to always use `[Symbol.observer]` when dealing with DOM events, in case for some reason that event is dispatched synchronously?

**Comment by zenparsing** [comment]
@blesh That's right.

> Meaning in the current design, it would be advisable to always use [Symbol.observer] when dealing with DOM events

Yeah, async subscription doesn't make any sense for DOM events.  But making people use `[Symbol.observer]` for this case is just *really* bad ergonomics.  Especially since there is no zalgo risk here.

Maybe you could say that DOM observables are a subclass of Observable with an override of `subscribe` which does subscription synchronously, but that seems really hacky to me.  

Also, it feels like bad API design to have two methods for subscribing.

**Comment by benjamingr** [comment]
> Maybe you could say that DOM observables are a subclass of Observable with an override of subscribe which does subscription synchronously, but that seems really hacky to me.

Also a huge violation of the liskov substitution principle as guarantees are not preserved.

The question here is whether or not synchronous DOM events are common or worth releasing Zalgo generally over.

**Comment by blesh** [comment]
>  If you'd like to see, I've posted an issue with async subscription in a code example above.

Thanks, I missed that, I'll have a look.

**Comment by blesh** [comment]
@benjamingr I'm too dumb to see the code sample you're speaking of above.

The only solution I can think of to support async subscription that allows sync dispatch to carry through would be to buffer values that were `next`-ed at the observer after subscribe was called but prior to actual subscription, then flush them after subscription. This seems, at it's surface at least, to be a bad idea. Do we really want to buffer N values prior to between subscription being called and actually occurring? Is that understandable/expected behavior? I'm unsure.

**Comment by benjamingr** [comment]
@blesh I'm referring to https://github.com/zenparsing/es-observable/issues/38#issuecomment-119199383

> The only solution I can think of to support async subscription that allows sync dispatch to carry through would be to buffer values that were next-ed at the observer after subscribe was called but prior to actual subscription, then flush them after subscription. This seems, at it's surface at least, to be a bad idea. Do we really want to buffer N values prior to between subscription being called and actually occurring? Is that understandable/expected behavior? I'm unsure.

This is really the crux. 

Work with me here, I'm trying to come up with use cases where buffering an unbounded amount of things synchronously is a bad idea and is a real use case. I think a use case where this is actually happening would go a _huge_ way for support of a sync subscription API and probably to make it ergonomic.

As you mentioned, the performance hit is only when values are dispatched synchronously to a synchronously attached `susbscribe` and were not being buffered anyway.

So far, I've been unable to come up with use cases where this sort of buffering is a problem.

**Comment by zenparsing** [comment]
@blesh 

I'm not proposing this as the actual solution, but here's another possible (and similar) solution:

- Make subscription always synchronous.
- Any notifications that occur while the subscriber function is being executed get queued up.  After the subscriber function finishes, if there are any queued notifications, then schedule a job to deliver them.

This buffering and scheduling behavior makes me nervous; it seems surprising and magical.

**Comment by benjamingr** [comment]
@zenparsing 

> This buffering and scheduling behavior makes me nervous; it seems surprising and magical.

Not to end users. End users would just see data run through the subscription asynchronously and know that regardless of where they got the observable from the order of execution is guaranteed.

**Comment by zenparsing** [comment]
Actually, my previous suggestion doesn't work all, because notifications could come in out-of-order.  A notification sent after the subscribe call but before "nextTick" would be received before the buffered notifications.  Bleh...

@jhusain Agreed, but it seems like with the current two-method setup consumers have to do the opt-in.  In the case of DOM events, we don't want users to have to explicitly say `[Symbol.observer]`.  How do you see that working?

**Comment by jhusain** [comment]
I'm not yet convinced that this is a prevalent enough scenario to engineer an ergonomic API around. It would be great to know just how many Event Targets synchronously notify. My guess is that this is not a very common hazard in practice.

**Comment by blesh** [comment]
> Not to end users. 

Queueing sync values for async subscription doesn't really have a precedent besides *maybe* Promises, but Promises aren't *subscribed* to they're just *resolved* synchronously and then *notify* async, they can't create and teardown their underlying data sources.

We could specify that Observable always notifies asynchronously, but subscribes synchronously, but then we're back in the same issue with EventTarget.

**Comment by blesh** [comment]
> It would be great to know just how many Event Targets synchronously notify. My guess is that this is not a very common hazard in practice.

@jhusain  integration tests come to mind. People write some component, and in their test they mount the component, call `click()` on some element they've selected, and then test for an outcome. I think I've seen this in a lot of places.

**Comment by trxcllnt** [comment]
@benjamingr 
>I'm not sure what this demonstrates at all except that changing schedulers changes behaviour.

Async subscribe === scheduling subscription on the trampolining scheduler, except since it isn't parameterized, it's impossible to achieve the depth-first tree walk behavior.

>appealing to 'pure functional programming' without any such data

The core features of Observables include full control over *when* code is executed, and deterministic memory management. Async subscribe irreversibly deprives the developer of those features. If you start with synchronous subscribe, you can opt-in to async subscribe. If you start with async subscribe, you can never opt back in to synchronous subscribe.

The misunderstanding here is that Observables aren't analogous to Promises, EventEmitters, or even Streams. Observables are immutable Functions that can return zero or more values whenever it wants (synchronously or asynchronously). They happen to be Objects because JavaScript doesn't include a `Function.create`, but the analogy holds: `Observable` :: `Function`, and `Observable.subscribe` :: `Function.call`.

You wouldn't expect `someFunction.call()` to invoke the function asynchronously, so why would you expect `someObservable.subscribe()` to invoke the Observable's `subscribe` asynchronously?

>monadic pure state book-keeping has proven to be very unpopular

I think the popularity of jQuery, underescore, and om are compelling counter-arguments.

> @zenparsing: In order to protect against zalgo, "well-behaved" observables should never execute any observer callbacks synchronously.
> @benjamingr: Here, if there is no zalgo protection, we have a race condition based on whether or not getTwo cached the result before. This is a major class of bugs people complain about and the reason people dislike sync subscription is because of that particular use.

Zalgo isn't an ancient monster in the bay and preys on unsuspecting startup workers in southbeach. Zalgo is a loan shark developers call when they're under time pressure and have to raise a bit of tech debt, but he breaks your knees when you don't pay him back on time. Zalgo's still in business because JavaScript doesn't provide the necessary language features to avoid calling him in the first place. This solution is like saying, "we need more knee surgeons to fix all the knees Zalgo breaks," instead of just throwing Zalgo in prison for loan sharking.

**Comment by zenparsing** [comment]
> We could specify that Observable always notifies asynchronously, but subscribes synchronously, but then we're back in the same issue with EventTarget.

@blesh Yes, and also that means observation would potentially require O(n) space for buffering.  We really would like observation to be O(1) space, given that it should be appropriate for delivering huge datasets.

**Comment by RangerMauve** [comment]
@blesh I kinda like the idea of subscribing being sync and notification being async. Any particular reason why that's a bad idea?

**Comment by blesh** [comment]
@zenparsing very true.

> If you start with synchronous subscribe, you can opt-in to async subscribe. If you start with async subscribe, you can never opt back in to synchronous subscribe.

@trxcllnt I agree with this wholeheartedly. 

> Observables are immutable Functions that can return zero or more values whenever it wants (synchronously or asynchronously).

This is correct. I think the TC39 needs to realize that Observables are a proposal more on-par with something like adding other functional constructs such as first-class functions, lambdas, Function.prototype.apply, or es7 function bind. Functional programming in JavaScript is on the rise, and this would add a powerful feature to the language.

**Comment by blesh** [comment]
@RangerMauve just as @zenparsing pointed out, you could easily end up spamming your micro-task queue with tons of observed values.  Think of `Observable.range(0, Number.MAX_VALUE)`.

**Comment by RangerMauve** [comment]
@blesh Would the "trampolining" method mentioned earlier help alleviate a lot of the overhead from that?

**Comment by benjamingr** [comment]
@zenparsing 

> Actually, my previous suggestion doesn't work all, because notifications could come in out-of-order. A notification sent after the subscribe call but before "nextTick" would be received before the buffered notifications. Bleh...

Why? The semantics of micro tasks vs macrotasks ensure that exactly this will not happen. The handlers will fire _after every microtask_ but _before any macrotask_ (the "late queue"). 


@blesh 


> This is correct. I think the TC39 needs to realize that Observables are a proposal more on-par with something like adding other functional constructs such as first-class functions, lambdas, Function.prototype.apply, or es7 function bind. 

That's very hard to argue given observables can, and have been implemented in user land successfully before. We need to show current APIs implemented in terms of observables like Object.observe, DOM APIs, Node APIs and so on and explain how observables solve the issue better. On my side it would be easier in terms of "no need to download RxJS and wrap APIs" but that's about it although this is a bit off topic. (To be clear though, I'm active here because I'm _very_ interested in seeing observables in the language and I think a _good default_ is incredibly important for users).

I think the two people who have been consistently talking in terms of functional programming features in the language are @domenic and @jhusain and I think it would be great if https://github.com/zenparsing/es-observable/issues/34 could get a little love. If we're talking FP I wondered if laziness is a property of the comonadability of observables that function-classes can encapsulate (like [this for promises for example](http://jsfiddle.net/r5aL327x/)) and me , @domenic and @headinthebox are (in the initial stages of) discussing this, I'm still figuring this out but I think function classes have a huge potential for the language and if we include them lots of nice things happen (`o.subscribe(sub)` just becomes `o(sub)` and a lot of other interesting things). That's a separate proposal though :)

**Comment by benjamingr** [comment]
@RangerMauve 

> @blesh Would the "trampolining" method mentioned earlier help alleviate a lot of the overhead from that?

No, if you do:

```
Observable.range(0, 10000).subscribe(function(v){
    console.log(v);
});
console.log("Here);
```

You would need to queue 10000 and pump them one by one. Observables are a very strict (non lazy) type in that regard.

**Comment by RangerMauve** [comment]
@benjamingr Ah, that makes sense. Thanks for the explanation.

**Comment by blesh** [comment]
> we're still figuring out if cases like this where 10000 items are created synchronously are common

I think I was talking with @trxcllnt and there are implementations of Animation written in Rx that actually do something like `Rx.Observable.range(0, Number.MAX_VALUE)` and then map/flatMap those into animation ticks and takeUntil the animation completes.

I'm not sure what the exact implementation is, or how common it is, so perhaps @trxcllnt could elaborate, but that's one I can think of off the top of my head that could get hairy with default async subscription or observation.

**Comment by benjamingr** [comment]
@blesh that would be fantastic, although keep in mind subscription would also need to be synchronous here.

**Comment by trxcllnt** [comment]
@benjamingr @zenparsing 

After further thought, I've identified two more operations that are impossible without synchronous subscription -- the grouping ([groupBy](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/groupby.md), [groupByUntil](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/groupjoin.md), [groupJoin](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/groupjoin.md)) and [window](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/window.md) operators.

### Description
These operators split an Observable of items into an Observable of inner Observables of items, based on custom identifying logic. When the operators are `next`'d a value, they may identify the need to create a new inner Observable. They `next` the new inner Observable to the subscriber, who has a chance to subscribe to it immediately. After `next`ing the new inner Observable, the operators `next` the original value to the new inner Observable.

This affects anybody who wants to group events by type, count, time, or relation.

### Why synchronous?
If subscription is asynchronous, the subscription to the inner Observable won't have been scheduled by the time the splitting operator `next`s the value, and the new inner Observer won't hear the event.

### Code
Example JSBin [here](http://jsbin.com/fupakateqe/edit?js,output).

```javascript
var Observable = Rx.Observable;
var currentThread = Rx.Scheduler.currentThread;
var RAFScheduler = Rx.Scheduler.requestAnimationFrame;
var canvas = document.getElementById('c');
var context = canvas.getContext('2d');

canvas.width = window.innerWidth;
canvas.height = window.innerHeight;
context.font = "12px Helvetica";
context.fillStyle = "white";

function randomNumbers(scheduler) {
  return Observable.create(function subscribe(observer) {
    scheduler.scheduleRecursive(function(reschedule) {
      observer.onNext(Math.random());
      reschedule();
    });
  });
}

randomNumbers(RAFScheduler)
  .map(function(value, allColorsCount) {
    if(value < 0.33) {
      return {
        color: "green",
        count: allColorsCount
      };
    } else if(value < 0.66) {
      return {
        color: "blue",
        count: allColorsCount
      };
    } else {
      return {
        color: "red",
        count: allColorsCount
      };
    }
  })
  .groupBy(function(x) { return x.color; })
  .flatMap(function(innerObservable, colorIndex) {
    return innerObservable.map(function(x, colorCount) {
      var percent = colorCount / x.count;
      var boxWidth = Math.round(canvas.width * 0.33);
      var boxHeight = Math.round(canvas.height * percent);
      var boxX = colorIndex * boxWidth;
      var boxY = canvas.height - boxHeight;
      var percentText = x.color + ": " + Math.round(percent * 10000)/100 + "%";
      var textMetrics = context.measureText(percentText);
      return {
        percent: percent,
        color: x.color,
        x: boxX,
        y: boxY,
        textX: boxX + Math.abs(boxWidth - textMetrics.width) * 0.5,
        textY: boxY - 12,
        percentText: percentText,
        width: boxWidth,
        height: boxHeight
      };
    });
  })
  .subscribe(function(box) {
    context.clearRect(box.x, 0, box.width, canvas.height);
    context.fillStyle = box.color;
    context.fillRect(box.x, box.y, box.width, box.height);
    context.fillStyle = "black";
    context.fillText(box.percentText,
                     Math.round(box.textX),
                     Math.round(box.textY));
  });
```

**Comment by trxcllnt** [comment]
> @benjamingr it would be great if #34 could get a little love

Yes! That proposal perfectly captures the essence of Observables, and beyond sweet syntactic sugar and potential closure optimizations, would eliminate any ambiguity around what an Observable actually is.

**Comment by blesh** [comment]
@trxcllnt I thought about mentioning this one, too. It seems like one could get around that by buffering in the outer GroupByObserver, but then you're going to run into the same "O(n)" buffering problem @zenparsing mentions above. You're completely right that async subscription would be way less than ideal.

**Comment by trxcllnt** [comment]
@blesh correct. If you schedule subscriptions, you have to schedule `next`'s through the `groupBy` Observable or risk dropping values. But then you don't have a "groupBy" operator, you have a "groupByAndObserverOn" operator.

**Comment by zenparsing** [comment]
@jhusain I'm not so sure that async subscription can survive this.  The general idea is:

Suppose that `subscribe` is called at time *t0*, and the subscription is scheduled to run at *tS*.  Then any event *tN* where *N* > 0 will be missed if *N* < *S*.

I think that covers the issues for both EventTarget and groupBy.  @trxcllnt do you agree?

**Comment by trxcllnt** [comment]
@zenparsing yes, I think that covers everything.

**Comment by benjamingr** [comment]
@trxcllnt the synchronous example in groupBy is excellent, but as @zenparsing and @blesh pointed out, it can be overcome by buffering. I do see the pain point in implementing a userland `.flatMap`-type operation though. I'd love to see benchmarks showing this is slow (we'll iterate them) as those would go a long way to justify sync subscription to the TC.

**Comment by domenic** [comment]
Just getting back to this 50-message-deep thread after a plane flight and night's sleep... Here are some things that I'm picking up on.

@trxcllnt

Setting the scheduler is not part of this proposal; if that is a prerequisite for safely using observables then that needs to be added to this proposal (which will presumably need to go back to stage 0 since that's a pretty big change).

@jhussein

If serving as a basis for EventTarget is an important part of this proposal, I think there's a lot more work to do there. The details on capture, bubbling, cancelation (which is not compatible with observable cancelation, note) need to be reproduced, as do the exact timing issues and ordering issues surrounding e.g. removing an event listener mid-capturing phase. I am not at all convinced that sync vs. async subscription is the blocker here, and I think it'd be premature to use EventTarget as a motivating use case for sync subscription before it's made clear that observable can handle the rest of the EventTarget semantics.

Also,

> Note that scheduling notification is also not an option, because preventDefault() must be invoked within the same job.

is not really true; preventDefault() can be called at any time. The question is when the UA checks whether the event is canceled. Click is a special case here; all other events dispatched from script do not have "default actions" that do such checks. See e.g. https://codereview.chromium.org/894913002/ and linked bugs/spec issues.

I personally think it would be a lot more feasible to have some kind of EventTarget -> observable adapter. However, others on the Chrome team are opposed to observables unless they can actually be used as part of the layered platform strategy, i.e., we can literally reimplement EventTarget on top of them and throw away existing EventTarget code. So that is a prerequisite for the proposal advancing, it is true.

@zenparsing

> This is what I'm thinking:
> - Subscription should happen synchronously, as pointed out above.
> - In order to protect against zalgo, "well-behaved" observables should never execute any observer callbacks synchronously.

This seems pretty reasonable to me on the face of it, and I found it very surprising that observables as-proposed went in a different direction than promises in this regard (which also do sync-subscribe, async-call-of-listener). I still have another 35 messages to work through but I am curious if this ends up going where I think it's going.

@benjamingr @jhussein

I/O actually *requires* unbounded buffering, and it is for this reason observables are not suitable for I/O.

@jhussein

> It would be great to know just how many Event Targets synchronously notify. My guess is that this is not a very common hazard in practice.

All of them do when used with dispatchEvent from author code. When UAs dispatch events they always enqueue a task to do so however.

@blesh

> I think the TC39 needs to realize that Observables are a proposal more on-par with something like adding other functional constructs such as first-class functions, lambdas, Function.prototype.apply, or es7 function bind

This attitude will make advancing observables through TC39 very difficult, I can assure you. You called this FUD earlier so maybe you are not interested in talking about this rationally but I thought I'd try one more time.

**Comment by benjamingr** [comment]
Hope you had a good flight @domenic :)

> Setting the scheduler is not part of this proposal; if that is a prerequisite for safely using observables then that needs to be added to this proposal (which will presumably need to go back to stage 0 since that's a pretty big change).

What @trxcllnt demonstrated there is that scheduling asynchronously vs. synchronously produces different execution order. As we later noted that was _not_ a motivating example for sync subscribe. 

> @jhusain If serving as a basis for EventTarget is an important part of this proposal, I think there's a lot more work to do there.

I definitely agree, I'd like to see a detailed survey and analysis of points of match and mismatch and I'd like to see more work with DOM people about compatibility. If EventTarget is indeed a goal of this proposal things need to be very clear.

As for `preventDefault`, other than click all other events should not be affected by synchronous dispatch as events will not be treated as trusted and the default action will not fire anyway. This is why I used `stopPropagation` in my example of the issue instead.

> @benjamingr @jhussein I/O actually requires unbounded buffering, and it is for this reason observables are not suitable for I/O.

Any form of I/O where the platform will have to wait requires bounded IO. It's really easy to show with:

    var streamOrEmitterOrAsyncIterator = url.readAsChunked("/petabyteFile.xml"); // because xmls are big :P
    while(true); 
    // consume and move forward

Pretty much, any non-preemptive model of async IO (and to an extent preemptive ones too) needs to deal with backpressure and buffering issues. The question is _how much additional buffering will be needed_ by making subscriptions async?

If `Observable.range(0, Infinity).subscribe(v => fn(v))` is code that is actually running then buffering is going to be problematic. In general if unbounded buffering is required for async subscription is a problem in _realistic use cases_ and _in addition to the regular buffering overhead_, then we need to consider sync subscription for that reason (and not sooner). 

So far we've yet to see code with this requirement but I think both @blesh and @trxcllnt are working on finding good examples. When and if they do, that would make a convincing argument for sync subscription in my humble opinion.

**Comment by domenic** [comment]
> As for preventDefault, other than click all other events should not be affected by synchronous dispatch as events will not be treated as trusted and the default action will not fire anyway. This is why I used stopPropagation in my example of the issue instead.

Right. The supposition there is that you'd want to interoperate between .addEventListener and .subscribe uses, which is not 100% clear to me. But I guess it would fall out naturally if you tried to the layering.

> Any form of I/O where the platform will have to wait requires bounded IO. It's really easy to show with:

This is probably off-topic but your example does not show this at all, as far as I can tell. I understand the rest of your post though so maybe it is immaterial. We could take it to another thread or IRC if you think it's important.

**Comment by blesh** [comment]
> You called this FUD earlier so maybe you are not interested in talking about this rationally but I thought I'd try one more time.

@domenic I apologize, and I deleted that comment. I have a strong impression you don't want this proposal to succeed, so I'm likely to misread every one of your issues and comments as having that underlying goal. 

That is a personal failing on my part that makes it hard to stay objective when reading your comments. 

Again, apologies. I'll try to stay objective.

**Comment by jhusain** [comment]
@domenic @slightlylate made it clear to me that Chrome wants to be able to build EventTarget on top of Observable. I'm trying to meet that bar, but the the basic type signature must stabilize first. @domenic do you acknowledge that sync subscription is required to build EventTarget? If Chrome plans to block the proposal if they can't build EventTarget on top of Observable, then I think we have to move forward with a sync subscription option and close this issue.

JH

**Comment by benjamingr** [comment]
This discussion seems to have died out - if anyone has anything to add let me know and I'll reopen.

**Comment by jhusain** [comment]
Reopening because as I'm putting together a summary of changes I'm somewhat confused by conclusion. It's clear we need sync subscription for EventTarget and for certain operators like GroupBy. It seems the original proposal of both an async and sync subscription option still works.

**Comment by zenparsing** [comment]
@jhusain we should open a new issue and reference this one, because this issue has gotten really long is so long.

#49

**Comment by arman-g** [comment]
In my case I do see one use case using it in angular interceptor to set a request header with bearer token if it is expired or does not exist in a local storage. The token request in the interceptor must be synchronous so that it can set the bearer token in the header before the async call starts.

**Comment by benjamingr** [comment]
@arman-g in your case you could subscribe synchronously since Angular interceptors are allowed to be async and there would be no benefit from synchronous subscription.

Not that it really matters that much since this proposal is mostly dead and I am sure most of us (at least me) are pretty ashamed of our conduct and the way we chose to converse here :] 

This conversation is a great example of seeking to be understood rather than to understand - it resulted in nothing, no one got smarter about whether or not firehosing or unbounded buffering (synchronous subscription) is important etc.
